taffy.py

```diff
@@ -0,0 +1,445 @@
+              "invert", "binvert", "bmotionvectorson", "wave_dots", "bwavedots", "wave_thick", "bwavethick"]:
+        preset_vars[s.lower()] = BOOL_RW
+
+    # read-only
+    for s in ["time", "bass", "mid", "treb", "bass_att", "mid_att", "treb_att", "progress"]:
+        preset_vars[s] = FLOAT_R
+    preset_vars['frame'] = INT_R
+    preset_vars['fps'] = INT_R
+    preset_vars['meshx'] = INT_R
+    preset_vars['meshy'] = INT_R
+
+    # matrix variables
+    for s in ["warp", "zoom", "rot", "zoomexp", "fzoomexponent", "cx", "cy", "dx", "dy", "sx", "sy"]:
+        preset_vars[s] = FLOAT_RW
+        matrix_vars[s] = FLOAT_RW
+    for s in ["x", "y", "ang", "rad"]:
+        matrix_vars[s] = FLOAT_R
+
+    for i in range(0, NUM_Q_VARIABLES):
+        preset_vars['q' + str(i)] = FLOAT_RW
+        matrix_vars['q' + str(i)] = FLOAT_RW
+
+    # shape
+    for s in ["r", "g", "b", "a", "border_r", "border_g", "border_b", "border_a", "r2", "g2", "b2", "a2", "x", "y",
+              "thickoutline", "enabled", "sides", "additive", "textured", "rad", "ang", "tex_zoom", "tex_ang",
+              ]:
+        shape_vars[s] = FLOAT_RW
+    for i in range(0, 8):
+        shape_vars['t' + str(i)] = FLOAT_RW
+    for i in range(0, NUM_Q_VARIABLES):
+        shape_vars['q' + str(i)] = FLOAT_RW
+    shape_vars['imageurl'] = STRING_RW
+    # TODO shape_points
+
+
+    # wave
+    for s in ["r", "g", "b", "a", "x", "y", "smoothing", "scaling"]:
+        wave_vars[s] = FLOAT_RW
+    for s in ["enabled", "bspectrum", "bdrawthick", "busedots", "badditive"]:
+        wave_vars[s] = BOOL_RW
+    for s in ["sep", "samples"]:
+        wave_vars[s] = INT_RW
+    for i in range(0, 8):
+        wave_vars['t' + str(i)] = FLOAT_RW
+    for i in range(0, NUM_Q_VARIABLES):
+        wave_vars['q' + str(i)] = FLOAT_RW
+
+
+    # TODO wave_points
+    # int [sample int_r
+    # float value1 float_r
+    # float value2 float_r
+
+
+class Shape:
+    def __init__(self, name):
+        self.name = str(name)
+        self.frame_eqns = []
+        self.point_eqns = []
+        self.variables = []
+        self.comments = []
+
+
+class Preset:
+    def __init__(self):
+        self.name = ''
+        self.variables = []
+        self.frame_init_eqns = []
+        self.frame_eqns = []
+        self.pixel_eqns = []
+        self.waves = []
+        self.shapes = []
+        self.comp_lang = 'hlsl'
+        self.comp = []
+        self.warp_lang = 'hlsl'
+        self.warp = []
+        self.comments = []
+
+    def get_shape(self, i):
+        i = int(i)
+        while len(self.shapes) <= int(i):
+            self.shapes.append(Shape(len(self.shapes)))
+        return self.shapes[i]
+
+    def get_wave(self, i):
+        i = int(i)
+        while len(self.waves) <= int(i):
+            self.waves.append(Shape(len(self.waves)))
+        return self.waves[i]
+
+
+# 'parse' right hand side of expression, allow substition of identifiers
+def scan_expr(expr, ident_callback):
```

Comment:
this is pretty wild.. you should look at https://kite.com/python/docs/pyparsing

Reply:
Thanks for feedback.  The identifier substitution was kind of a experiment, I might just remove it.  Thanks for the pyparsing suggestion.